### PR TITLE
[FrameworkBundle] [TwigBundle] Move the hinclude key away from templating

### PR DESCRIPTION
**Pull request by Simperfit**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | yes 
| Tests pass?   | yes    
| Fixed tickets | #30874 
| License       | MIT
| Doc PR        | to do when pr is validated.

Maybe I shouldn't move directly the config key from templating to the other, but since the templating component has been deprecated we may change this directly without deprecating that key alone, WDYT ?

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 1]
```diff
@@ -191,7 +191,8 @@ private function addFragmentsSection(ArrayNodeDefinition $rootNode)
                     ->info('fragments configuration')
                     ->canBeEnabled()
                     ->children()
-                        ->scalarNode('path')->defaultValue('/_fragment')->end()
+                    ->scalarNode('hinclude_default_template')->defaultNull()->end()
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: indentation should be kept

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -610,7 +611,6 @@ private function addTemplatingSection(ArrayNodeDefinition $rootNode)
                         ->then(function () { return ['enabled' => false, 'engines' => false]; })
                     ->end()
                     ->children()
-                        ->scalarNode('hinclude_default_template')->defaultNull()->end()
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Shouldn't we keep it and move it to the new option (with deprecation)?

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 2 above)
Comment (stof) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: We do. Otherwise, that's a BC break.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -487,7 +487,13 @@ private function registerFragmentsConfiguration(array $config, ContainerBuilder
 
             return;
         }
+        $container->setParameter('fragment.renderer.hinclude.global_template', $config['hinclude_default_template']);
 
+        if (!$container->hasParameter('templating.engines') || ['twig'] == $container->getParameter('templating.engines')) {
+            $container->getDefinition('fragment.renderer.hinclude')
+                ->addTag('kernel.fragment_renderer', ['alias' => 'hinclude'])
+                ->replaceArgument(0, new Reference('twig'));
```
Comment (stof) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: FramewokrBundle injecting a reference to `twig` is wrong, as it does not depend on TwigBundle. That code should stay where it was in TwigBundle IMO.

**Comment by Simperfit** [comment]
done @stof @fabpot

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -603,6 +604,16 @@ private function addTemplatingSection(ArrayNodeDefinition $rootNode)
         $rootNode
             ->children()
                 ->arrayNode('templating')
+                ->validate()
+                    ->ifTrue(function ($v) {
+                        return null !== $v['hinclude_default_template'];
+                    })
+                    ->then(function ($v) {
+                        @trigger_error('Using the key "hinclude_default_template" in the templating is deprecated since Symfony 4.3. Use the config key under fragments instead.', E_USER_DEPRECATED);
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: any reason we don't use the setDeprecated() method instead?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 5]
```diff
@@ -612,6 +623,7 @@ private function addTemplatingSection(ArrayNodeDefinition $rootNode)
                     ->children()
                         ->scalarNode('hinclude_default_template')->defaultNull()->end()
                         ->scalarNode('cache')->end()
+
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: unrelated

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 4 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: no reason, will change this

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -487,6 +487,7 @@ private function registerFragmentsConfiguration(array $config, ContainerBuilder
 
             return;
         }
+        $container->setParameter('fragment.renderer.hinclude.global_template', $config['hinclude_default_template']);
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Does it work even when providing the value via the old key and not via the new one?

**Comment by Simperfit** [comment]
It does work properly since I didnt removed anything I just added the
deprecation

Le dim. 7 avr. 2019 à 18:53, Fabien Potencier <notifications@github.com> a
écrit :

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 6 above)
Comment (chalasr) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: @Simperfit You are setting a parameter with no prior existence/emptiness check at 2 different places so only one can win (and potentially not the expected one). We need a test I guess :)

**Comment by Simperfit** [comment]
Ooow I understand, I will write one !

Le dim. 7 avr. 2019 à 21:56, Robin Chalas <notifications@github.com> a
écrit :

**Comment by Simperfit** [comment]
I've added an existence check to not replace if it has been already setted.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 7]
```diff
@@ -905,7 +909,9 @@ private function registerTemplatingConfiguration(array $config, ContainerBuilder
     {
         $loader->load('templating.xml');
 
-        $container->setParameter('fragment.renderer.hinclude.global_template', $config['hinclude_default_template']);
+        if (!$container->hasParameter('fragment.renderer.hinclude.global_template')) {
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: If will never be set as `registerTemplatingConfiguration` is called before `registerFragmentsConfiguration`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 8]
```diff
@@ -488,6 +488,10 @@ private function registerFragmentsConfiguration(array $config, ContainerBuilder
             return;
         }
 
+        if (!$container->hasParameter('fragment.renderer.hinclude.global_template')) {
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: This depends on the order of the calls between `registerFragmentsConfiguration` and `registerTemplatingConfiguration`. Here, we want the configuration on `fragments` to override the one on `templating`, but that's currently the other way around.

I would do it differently. If it's set in both configuration, throw an exception (as it means that the configuration is ambiguous).

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 8 above)
Comment (Simperfit) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: I understand now, I didn't have that in mind, thanks for the clarification, I'm going to modify in that way. Thanks for the review !

**Comment by Simperfit** [comment]
@fabpot Should be ok now, I've added a test.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 9]
```diff
@@ -906,7 +911,6 @@ private function registerTemplatingConfiguration(array $config, ContainerBuilder
         $loader->load('templating.xml');
 
         $container->setParameter('fragment.renderer.hinclude.global_template', $config['hinclude_default_template']);
-
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: should be reverted

**Comment by Simperfit** [comment]
@fabpot PR ready, CI failure not related to this PR.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 10]
```diff
@@ -487,6 +487,11 @@ private function registerFragmentsConfiguration(array $config, ContainerBuilder
 
             return;
         }
+        if ($container->hasParameter('fragment.renderer.hinclude.global_template') && (null !== $container->getParameter('fragment.renderer.hinclude.global_template') && null !== $config['hinclude_default_template'])) {
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: extra parentheses can be removed

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 11]
```diff
@@ -487,6 +487,11 @@ private function registerFragmentsConfiguration(array $config, ContainerBuilder
...
+            throw new \LogicException('Ambiguous configuration detected, prior to Symfony 4.3 you should not use both config keys ("templating.hinclude_default_template" and "fragments.hinclude_default_template", please use "fragments.hinclude_default_template".');
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: `You cannot set both "templating.hinclude_default_template" and "fragments.hinclude_default_template", please only use "fragments.hinclude_default_template".`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 12]
```diff
@@ -610,7 +611,7 @@ private function addTemplatingSection(ArrayNodeDefinition $rootNode)
                         ->then(function () { return ['enabled' => false, 'engines' => false]; })
                     ->end()
                     ->children()
-                        ->scalarNode('hinclude_default_template')->defaultNull()->end()
+                        ->scalarNode('hinclude_default_template')->setDeprecated('Using the key "hinclude_default_template" in the templating is deprecated since Symfony 4.3. Use the config key under fragments instead.')->defaultNull()->end()
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: `Setting "templating.hinclude_default_template" is deprecated since Symfony 4.3, use "fragments.fragments.hinclude_default_template" instead.`

**Review by fabpot — CHANGES_REQUESTED** [review]

**Comment by Simperfit** [comment]
done @fabpot

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 13]
```diff
@@ -610,7 +611,7 @@ private function addTemplatingSection(ArrayNodeDefinition $rootNode)
                         ->then(function () { return ['enabled' => false, 'engines' => false]; })
                     ->end()
                     ->children()
-                        ->scalarNode('hinclude_default_template')->defaultNull()->end()
+                        ->scalarNode('hinclude_default_template')->setDeprecated('Setting "templating.hinclude_default_template" is deprecated since Symfony 4.3, use "fragments.fragments.hinclude_default_template" instead.')->defaultNull()->end()
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: double fragments.fragments.

**Comment by Simperfit** [comment]
done @fabpot

**Review by fabpot — APPROVED** [review]
